Some websocket code cleanup

vortigont opened the pull request:
a bit of tidy work
add explicit constructors
remove extra calls, etc...

run checked, gives a tiny bit of reduction in code size for my project
```
// this PR 
Flash: [========  ]  82.5% (used 1460245 bytes from 1769472 bytes)     
// orig
Flash: [========  ]  82.5% (used 1460669 bytes from 1769472 bytes)
```

proddy commented:
If you're really trying to reduce Flash on an ESP32 try compiling with 
`platform = https://github.com/tasmota/platform-espressif32/releases/download/2024.05.00/platform-espressif32.zip`

It's a Tasmota special build of the Arduino library, based on Core 2.0.17 and IDF 4.4.7. It has some libs stopped out like mbedtsl, but has added support for IPv6 for example. It will save you 500KB in Flash and also give you a nice bump up on available heap space.

@mathieucarbou

mathieucarbou commented on src/AsyncWebSocket.h:
```diff
@@ -54,6 +54,8 @@
 #define DEFAULT_MAX_WS_CLIENTS 4
 #endif
 
+using ws_buffer_t = std::shared_ptr<std::vector<uint8_t>>;
```
I'm not a fan of using `using` keyword like that in a header in a global scope. This can lead to naming conflicts because it isn't scoped. But I will gladly merge the order changed. Thanks!

mathieucarbou commented:
> If you're really trying to reduce Flash on an ESP32 try compiling with `platform = https://github.com/tasmota/platform-espressif32/releases/download/2024.05.00/platform-espressif32.zip`
> 
> It's a Tasmota special build of the Arduino library, based on Core 2.0.17 and IDF 4.4.7. It has some libs stopped out like mbedtsl, but has added support for IPv6 for example. It will save you 500KB in Flash and also give you a nice bump up on available heap space.
> 
> @mathieucarbou



> If you're really trying to reduce Flash on an ESP32 try compiling with `platform = https://github.com/tasmota/platform-espressif32/releases/download/2024.05.00/platform-espressif32.zip`
> 
> It's a Tasmota special build of the Arduino library, based on Core 2.0.17 and IDF 4.4.7. It has some libs stopped out like mbedtsl, but has added support for IPv6 for example. It will save you 500KB in Flash and also give you a nice bump up on available heap space.
> 
> @mathieucarbou

thanks! I knew this option but not about the size decrease. I will much more try to compile compare with the most recent one on ESP-IDF 5.

`https://github.com/tasmota/platform-espressif32/releases/download/2024.06.10/platform-espressif32.zip`

 From what I heard from Jason, Tasmota is stripping down and removing some features not needed in Tasmota but I am not able to find a clear doc stating what's the difference.

mathieucarbou commented:
OK it was fast lol... They removed the ESP-IDF mqtt client which I am relying on...

```
    7 | #include <mqtt_client.h>
      |          ^~~~~~~~~~~~~~~
compilation terminated.
*** [.pio/build/trial-esp32-debug/src/init/Logging.cpp.o] Error 1
*** [.pio/build/trial-esp32-debug/src/init/Events.cpp.o] Error 1
In file included from include/YaSolR.h:21,
                 from src/init/MQTT.cpp:5:
.pio/libdeps/trial-esp32-debug/MycilaMQTT/src/MycilaMQTT.h:7:10: fatal error: mqtt_client.h: No such file or directory
```

So not an option in my projects.

mathieucarbou commented:
@vortigont : please let me know if you want to edit your PR or I'll do it to rollback the change regarding `using`.

vortigont commented:
> @vortigont : please let me know if you want to edit your PR or I'll do it to rollback the change regarding `using`.

Sure, I'll remove it, not a big deal here indeed.
Actually what I needed was  the `const` in `AsyncWebSocket::makeBuffer(const uint8_t * data, size_t size)`, the rest is what had caught my eye.

But let's discuss :)
There is not harm in aliasing with using if alias name is chosen carefully. Sometimes it might be quite feasible solution for seamless stitching user code with API expansion.
What comes to my mind you can change `std::shared_ptr<std::vector<uint8_t>>` into `std::vector<>` but with custom allocator for boards having PSRAM chip. And user code using `ws_buffer_t` type alias won't notice a thing.
It's just an example...

mathieucarbou commented:
> if alias name is chosen carefully.

Completely agree :-)

If the chosen name was more something in the spirit of this library, like `AsyncWebSocketSharedBuffer` or something similar, I wouldn't have put this comment.

If you want to change it for `AsyncWebSocketSharedBuffer` or a better name in the same spirit of the project classes, I am 100% for it 👍 I also do not like seeing that in the signatures (`std::shared_ptr<std::vector<uint8_t>>`) this is long and more like internals to me, internals that I've already hidden through the backward compatible `AsyncWebSocketMessageBuffer`, but kept part as the public API for backward compatibility.

**Maintain backward compatibility regarding the existing API is really what matters the most :-)**

mathieucarbou commented on src/AsyncWebSocket.cpp:
```diff
@@ -132,17 +132,11 @@ size_t webSocketSendFrame(AsyncClient *client, bool final, uint8_t opcode, bool
     : _buffer(std::make_shared<std::vector<uint8_t>>(size)) 
 {
     if (_buffer->capacity() < size) {
-        _buffer.reset();
-        _buffer = std::make_shared<std::vector<uint8_t>>(0);
```
The code here was right: it was coming from the original fork. 

This is a safety check in case the buffer cannot be allocated in `_buffer(std::make_shared<std::vector<uint8_t>>(size))` because of not enough heap, then we check that and we reset the buffer to a 0 size to free memory.

The original code from me-no-dev was also doing something similar when creating a `AsyncWebSocketMessageBuffer`.

The logic here: we ask to allocate a `AsyncWebSocketMessageBuffer` buffer of a specific size. If we cannot, then `AsyncWebSocketMessageBuffer` is not allocated at all.

vortigont replied:
that code seems all strange to me,
if, for some reason, `std::vector` won't be able to allocate it will throw `std::bad_alloc`, which in turn will propagate to `std::make_shared`, it won't create half allocated vector where you could check `_buffer->capacity()`, the code will crash anyway.
Also `_buffer.reset();` does not reset vector's capacity, it will release the shared pointer from the object, but that does not make sense if you are going to create a new empty one instead. Just the one that is already there.

Actually the should not be anything at all, if the allocation was successful, then you are good. If not, then code will crash anyway, either in `throw`  or if some miracle happens, later it will corrupt heap when user code start writing in vector's empty data* pointer.

mathieucarbou replied:
Thanks for the explanations! I've misread when using it and did't see the doc link to the allocator saying it will throw. My mistake. So all is good! Thanks for the PR!

mathieucarbou commented on src/AsyncWebSocket.cpp:
```diff
@@ -152,16 +146,10 @@ AsyncWebSocketMessageBuffer::AsyncWebSocketMessageBuffer(size_t size)
     : _buffer(std::make_shared<std::vector<uint8_t>>(size)) 
 {
     if (_buffer->capacity() < size) {
-        _buffer.reset();
-        _buffer = std::make_shared<std::vector<uint8_t>>(0);
```
Same comment as before - the code was right.

mathieucarbou commented on src/AsyncWebSocket.cpp:
```diff
@@ -1309,7 +1298,8 @@ AsyncWebSocketResponse::AsyncWebSocketResponse(const String& key, AsyncWebSocket
     sha1(key + WS_STR_UUID, hash);
 #else
-    String k = key + WS_STR_UUID;
+    String k(key);
+    k += WS_STR_UUID;
```
Can you please change to that to avoid reallocating and internal buffer copy:
```c++
#define WS_STR_UUID_LEN 36

[...]

String k;
k.reserve(key.length() + WS_STR_UUID_LEN)
k += key;
k += WS_STR_UUID;
```

Pablo2048 replied:
If we are preserved the target string the better way to concatenate is:
```cpp
k.concat(key);
k.concat(WS_STR_UUID);
```
this way is more memory allocation friendly than ordinary `k += ...` IMHO...

mathieucarbou replied:
@Pablo2048 : please correct me if I'm wrong, but `+=` of `String` is implemented as:
```c++
        String & operator +=(const char *cstr) {
            concat(cstr);
            return (*this);
        }
```
So it is doing a concat and returns the same reference. I think the implementors made it in a way to avoid a copy.

Pablo2048 replied:
:D well you got me. I didn't expect that the += operator is overloaded. You are right, in this use it's the same in view of memory allocation...

vortigont commented:
OK, let me do some fixes :)

BTW, @mathieucarbou is this fork still 8266 compatible? It frustrates me when stumbling across 8266's PROGMEMS's :)

mathieucarbou commented:
> OK, let me do some fixes :)
> 
> BTW, @mathieucarbou is this fork still 8266 compatible? It frustrates me when stumbling across 8266's PROGMEMS's :)

Yes it is ! Finally I've made sure to keep it, and include a CI for 8266= ;-)

vortigont commented:
> Yes it is ! Finally I've made sure to keep it

WoW! That's a challenge! I gave up all 8266 projects, no more time waste on it )

So, let it be `AsyncWebSocketSharedBuffer` then :)
Updated with what we've discussed. Also moved 50 bytes buffer alloc on stack, guess that's not a big deal even on 8266.

mathieucarbou commented:
> WoW! That's a challenge! I gave up all 8266 projects, no more time waste on it )

Actually this is just maintenance, not feature dev.  So that's ok, no overhead ;-)

> Also moved 50 bytes buffer alloc on stack

Yes for the SHA1, not a big deal 👍

proddy commented:
> > If you're really trying to reduce Flash on an ESP32 try compiling with `platform = https://github.com/tasmota/platform-espressif32/releases/download/2024.05.00/platform-espressif32.zip`
> > It's a Tasmota special build of the Arduino library, based on Core 2.0.17 and IDF 4.4.7. It has some libs stopped out like mbedtsl, but has added support for IPv6 for example. It will save you 500KB in Flash and also give you a nice bump up on available heap space.
> > @mathieucarbou
> 
> > If you're really trying to reduce Flash on an ESP32 try compiling with `platform = https://github.com/tasmota/platform-espressif32/releases/download/2024.05.00/platform-espressif32.zip`
> > It's a Tasmota special build of the Arduino library, based on Core 2.0.17 and IDF 4.4.7. It has some libs stopped out like mbedtsl, but has added support for IPv6 for example. It will save you 500KB in Flash and also give you a nice bump up on available heap space.
> > @mathieucarbou
> 
> thanks! I knew this option but not about the size decrease. I will much more try to compile compare with the most recent one on ESP-IDF 5.
> 
> `https://github.com/tasmota/platform-espressif32/releases/download/2024.06.10/platform-espressif32.zip`
> 
> From what I heard from Jason, Tasmota is stripping down and removing some features not needed in Tasmota but I am not able to find a clear doc stating what's the difference.

Yes I'm using "platform = https://github.com/tasmota/platform-espressif32/releases/download/2024.06.10/platform-espressif32.zip" which is Arduino 3.0.1 and it's working well.

What I see Jason removed from his platform builds were mbedtls and all the cyphers except the ones used to connect to the WiFi. For SSL it uses BearSSL which has a smaller footprint and outperforms mbedtls, similar to what you did recently. Also the SPIFFS and Arduino NIMBLE has been removed and replaced with h2zero NimBLE (esp-nimble-cpp). https://github.com/Jason2866/esp32-arduino-lib-builder

vortigont commented:
BTW, I'm curious why mbedtls is so heavy, due to included certs bundle or it's a code footprint?
As I remember, mbedtls is using HW accelerated crypto on ESP32, and so does WolfSSL.
But I'm not sure about BearSSL.

mathieucarbou commented:
I don't know either. I've tried to use it on Arduino 3 to connect to a secured MQTT server, and I had to set the server certificate (using the ESP-IDF mqtt client). I found no way to connect insecurely to any secured server, and I found no way to eventually re-use a certificate database that would contain certificate authorities.
